File: vol1/java-examples/src/main/java/com/heatonresearch/aifh/examples/randomize/EvaluatePI.java
/*
 * Artificial Intelligence for Humans
 * Volume 1: Fundamental Algorithms
 * Java Version
 * http://www.aifh.org
 * http://www.jeffheaton.com
 *
 * Code repository:
 * https://github.com/jeffheaton/aifh

 * Copyright 2013 by Jeff Heaton
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * For more information on Heaton Research copyrights, licenses
 * and trademarks visit:
 * http://www.heatonresearch.com/copyright
 */

package com.heatonresearch.aifh.examples.randomize;

import com.heatonresearch.aifh.randomize.GenerateRandom;
import com.heatonresearch.aifh.randomize.MersenneTwisterGenerateRandom;

/**
 * Approximate PI by Monte Carlo.
 * <p/>
 * http://en.wikipedia.org/wiki/Monte_Carlo_method
 */
public class EvaluatePI {

    /**
     * Random number generator.
     */
    private final GenerateRandom rnd = new MersenneTwisterGenerateRandom();

    public void process() {
        long tries = 0;
        int success = 0;
        int lastUpdate = 0;

        double x, y;

        for (int i = 0; i < 1000000000; i++) {
            // pick a point at random.
            x = rnd.nextDouble();
            y = rnd.nextDouble();

            tries++;

            // was the point inside of a circle?
            if (x * x + y * y <= 1)
                success++;

            lastUpdate++;
            if (lastUpdate >= 1000000) {
                final double pi = 4 * (double) success / (double) tries;
                System.out.println("Tries=" + tries + ", pi=" + pi);
                lastUpdate = 0;
            }

        }
    }

    /**
     * The main method.
     *
     * @param args Not used.
     */
    public static void main(final String[] args) {
        final EvaluatePI program = new EvaluatePI();
        program.process();
    }
}

File: vol1/csharp-examples/AIFH-Vol1/Examples/Randomize/EvaluatePI.cs
using System;
using AIFH_Vol1.Core.Randomize;

namespace AIFH_Vol1.Examples.Randomize
{
    /// <summary>
    /// Approximate PI by Monte Carlo.
    /// 
    /// http://en.wikipedia.org/wiki/Monte_Carlo_method
    /// </summary>
    public class EvaluatePI
    {
        /// <summary>
        /// The name of this example.
        /// </summary>
        public static string ExampleName = "Approximate PI by Monte Carlo";

        /// <summary>
        /// The chapter this example is from.
        /// </summary>
        public static int ExampleChapter = 4;

        /// <summary>
        /// Random number generator.
        /// </summary>
        private IGenerateRandom _rnd = new MersenneTwisterGenerateRandom();

        public void Process()
        {
            long tries = 0;
            int success = 0;
            int lastUpdate = 0;

            for (int i = 0; i < 1000000000; i++)
            {
                // pick a point at random.
                double x = _rnd.NextDouble();
                double y = _rnd.NextDouble();

                tries++;

                // was the point inside of a circle?
                if (x * x + y * y <= 1)
                    success++;

                lastUpdate++;
                if (lastUpdate >= 1000000)
                {
                    double pi = 4 * (double)success / tries;
                    Console.WriteLine("Tries=" + tries + ", pi=" + pi);
                    lastUpdate = 0;
                }

            }
        }

        /// <summary>
        ///     The entry point for this example.  If you would like to make this example
        ///     stand alone, then add to its own project and rename to Main.
        /// </summary>
        /// <param name="args">Not used.</param>
        public static void ExampleMain(string[] args)
        {
            var program = new EvaluatePI();
            program.Process();
        }
    }
}
